single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -0,0 +1,875 @@
+                logger.error(early_removal_message)
+                raise EarlyRemovalOfConfigServerError(early_removal_message)
+            if self.state.is_role(MongoDBRoles.SHARD) and self.state.shard_relation is not None:
+                logger.info("Wait for shard to drain before detaching storage.")
+                self.charm.status_manager.to_maintenance("Draining shard from cluster")
+                mongos_hosts = self.state.shard_state.mongos_hosts
+                self.shard_manager.wait_for_draining(mongos_hosts)
+                logger.info("Shard successfully drained storage.")
+            return
+
+        try:
+            # retries over a period of 10 minutes in an attempt to resolve race conditions it is
+            # not possible to defer in storage detached.
+            logger.debug("Removing %s from replica set", self.state.unit_peer_data.internal_address)
+            for attempt in Retrying(
+                stop=stop_after_attempt(600),
+                wait=wait_fixed(1),
+                reraise=True,
+            ):
+                with attempt:
+                    # remove_replset_member retries for 60 seconds
+                    self.mongo_manager.remove_replset_member()
+        except NotReadyError:
+            logger.info(
+                "Failed to remove %s from replica set, another member is syncing",
+                self.charm.unit.name,
+            )
+        except PyMongoError as e:
+            logger.error("Failed to remove %s from replica set, error=%r", self.charm.unit.name, e)
+
+    @override
+    def on_update_status(self) -> None:
+        """Status update Handler."""
+        if not self.pass_status_basic_checks():
+            return
+
+        if self.state.is_role(MongoDBRoles.SHARD):
+            shard_has_tls, config_server_has_tls = self.shard_manager.tls_status()
+            if config_server_has_tls and not shard_has_tls:
+                self.charm.status_manager.to_blocked("Shard requires TLS to be enabled")
+                return
+
+        if not self.mongo_manager.mongod_ready():
+            self.charm.status_manager.to_waiting("Waiting for MongoDB to start")
+            return
+
+        if self.substrate == Substrates.K8S:
+            self.upgrade_manager._reconcile_upgrade()
+
+        # It's useless to try to perform self healing if upgrade is in progress
+        # as the handlers would raise an UpgradeInProgressError anyway so
+        # better skip it when possible.
+        if not self.state.upgrade_in_progress:
+            try:
+                self.perform_self_healing()
+            except ServerSelectionTimeoutError as e:
+                logger.info(f"Failed to perform self healing: {e}")
+
+        self.charm.status_manager.process_and_share_statuses()
+
+    def on_set_password_action(self, username: str, password: str | None = None) -> tuple[str, str]:
+        """Handler for the set password action."""
+        user = get_user_from_username(username)
```

Comment:
missing the call to `assert_pass_password_checks `

Reply:
The call is done on the events side.
But you're right and I'll bring it here!